### DISCUSSION TITLE
Most efficient way to create 7,777 unique overlays of multiple gifs?

### DISCUSSION THREAD

**Discussion by mugungalabz** [ImageMagick/ImageMagick#4346] [Consulting]
Hello, I need the fastest most cpu efficient way to generate 7,777 unique variations of 1100x1100 px gifs.  More than happy to pay an expert who wants to consult me through this on the side. I will have 7 png backgrounds, and about 50 different alpha animations to layer on top. So each individual gif will be unique and have 1 png plus 4-8 of the animations.

The inputs can be either a gif form (single layer alpha gif animations) or a series of png frames. 

My two current tactics are: 

1) Just combine gifs like this, seems to be doing what I need, however this is moving too slow for me 4/5 minutes per .gif and I need like 1 minute. Would unoptimized but giant gifs actually move faster here?

`        "convert " + "\( resources/bb/background.png \) " + "null: " +
        "\( output/anim01.gif -layers coalesce \) " +
        "-gravity center -layers composite null: " +
        "\( output/anim02.gif -layers coalesce \) " +
        "-gravity center -layers composite " +
        "output/multitest.gif" `

2) Combining all the pngs for each frame FIRST, then creating a .GIF at the end. 

`composite background.png layer01frame01.png layer02frame01.png layer23frame01.png final01.png`
`composite background.png layer01frame02.png layer02frame02.png layer23frame02.png final02.png`
....and so forth...then create a gif. 

I could really use some help. New to Image Magick so I don't have the time or experience, but this seems like the correct tool for what I'm doing. I am wrapping everything in python scripts so all I need is the best image magick syntax.

Thanks in advance.

**Comment by snibgo** [comment]
What version of IM? If v7, I suggest you use "magick" instead of "convert", and "magick composite" instead of "composite".

How many frames in each animated GIF? 50?

You have 7 different input PNGs. How many input GIFs?

How many pixels in each PNG background and each GIF frame? Are they all 1100x1100 pixels?

Your first command looks as if it should work. But you will be using the same PNGs for multiple commands (about one thousand times), so I suggest first write each one to a MIFF or MPC file. Then run your command multiple times reading the MIFF or MPC instead of the PNG. This will be much faster.

The same applies to the input GIFs.

MIFF or MPC? In theory, MPC should be faster for write-once read-many, but I often find that MIFF is faster.

**Reply by mugungalabz** [reply]
Version 7. 

Unfortunately we are talking 300-400 frames, not 100% finalized. 

Each gif is 1 png + 4 to 8 different gifs, chosen from a pool of about 7 pngs/50 gifs. 

everything can be standardized to 1100px by 1100px by the gaphic designer, so I'm trying to talk them into the smallest pixel size they'll go with. All inputs will be uniform size, though. 

I'm unfamiliar with MIFF and MPC, and how they relate here. I can do some light(?) reading on that.

**Comment by snibgo** [comment]
Each GIF is, say, 400 frames each 1100*1100 pixels. This is 484 million pixels. If you have three channels of 4 bytes each, this would be 5.8 GB of memory. Your first command needs two of these in memory at any one time, so that needs 11.6 GB of memory. If your computer is big enough (say, 32 GB memory) this won't be a problem.

I don't think HDRI would be useful for you, so your build of IM could be integer-only Q16, so only 2 bytes per channel per pixel, halving memory use. This is also faster than HDRI (which uses floating-point).

MIFF and MPC are file formats, similar to GIF etc, but they are internal to ImageMagick (IM). They are much faster to read and write than PNG or GIF. See the brief descriptions in https://www.imagemagick.org/script/formats.php

To create MIFF files from PNG and GIF formats:
```
magick myfile.png myfile.miff

magick abcd.gif -layers coalesce abcd.miff
```
Expect the MIFF or MPC files to be larger than PNG or GIF files.

**Reply by mugungalabz** [reply]
Cool, I converted all the raws to Miffs instead of gifs, and combined those into gifs. It is running smoother and about twice as fast. I have another question now. Some of the gifs are different lengths, I want the final frame of the shorter gifs to freeze in place so that all the animations stack. Once they've all finished, I want the final frame to hold for a few seconds, so either fill in redundant frames, or set the duration of the final frame to 3000 milliseconds.

Super Bonus: If I can specify which frame to start each layer on, even better, so maybe one animation starts 3 seconds in on the 75th frame, and another stars 5 seconds in on the 125th frame. 

Here's where I'm at. 

`magick *.png out.miff.`

Then used those for inputs for the gif:

```
magick -delay 4 -loop 0
\( background.miff \) null: 
 \(  layer1.miff -layers coalesce \) 
   -layers composite null: 
 \(  layer2.miff -layers coalesce \) 
   -layers composite -layers optimize
output.gif
```

**Comment by snibgo** [comment]
In your command, you have:
```
 \(  layer1.miff -layers coalesce \)
```
If you created layer1.miff with "-layers coalesce", then you don't need to do that again in your main command.

> Some of the gifs are different lengths, I want the final frame of the shorter gifs to freeze in place so that all the animations stack.

So you need to find the number of frames in each input gif, and find the largest of these, so the shorter gifs can be extended. I suppose you want the output gif to have the same delay on every frame, except possibly the final output frame. So extending the shorter gifs would be by duplicating the final frame (or the initial frame, or something fancy).

So if layer1.miff needs to be extended by 20 frames, the main command could be:
```
magick -delay 4 -loop 0
\( background.miff \) null: 
 \(  layer1.miff -duplicate 20 \) 
   -layers composite null: 
 \(  layer2.miff -duplicate 0 \) 
   -layers composite -layers optimize
 -layers optimize
output.gif
```
You would calculate the numbers "20" and "0" in your Python script. I have added "-layers optimize" to try to reduce the size (in bytes) of the output file.

If you want the final output frame to stay on screen for longer, then that frame needs a longer delay. Before writing the output gif, insert this:
```
\( +clone -set delay 3000 \) -delete -2
```
This starts a new (empty) list of images, copies the last frame into that list, sets the delay on that frame, closes the list so that copied frame is added to the end of the main image list. So the last frame now has two copies, and only the second one has the long delay. We delete the second from last frame to remove the first of these copies.

> Super Bonus: If I can specify which frame to start each layer on, even better, so maybe one animation starts 3 seconds in on the 75th frame, and another stars 5 seconds in on the 125th frame.

No problem. After reading layer1.miff etc, "-delete 0-N" will remove the first N+1 frames from the image list. The number of frames to duplicate should take this into account.

**Reply by mugungalabz** [reply]
Thanks!  with adding -layer optimize at the end, my .gif is 1 MB as oppose to 100 Mb, and renders much faster, which is great. However the animation becomes a choppy and stuttery, and seems to actually play slower.  For context, it is moving animation so some of the pixels are returning to transparency. Basically a cleared frame animation: https://legacy.imagemagick.org/Usage/anim_basics/#cleared

Using -layers optimizeFrame or -layers optimizePlus was also choppy.

When I open those two versions on photoshop, I can see the large smooth .gif has an entire image fo each frame, whereas the small file I can see all the alpha layers its trying to stack up. Do you know of a better way to optimize the cleared frame animation so I can still shorten processing time and filesize but not have it as choppy? 

```
magick -delay 4 -loop 0
\( background.miff \) null: 
 \(  layer1.miff -duplicate 20 \) 
   -layers composite null: 
 \(  layer2.miff -duplicate 0 \) 
   -layers composite -layers optimize
output.gif
```

The duplicate frame worked for extending the last frame. How does it know it's the last frame of that layer to duplicate?   

I'm still working through -clone and -delete syntax.

**Reply by mugungalabz** [reply]
Ok, I spoke too soon and just found -layers optimizeTransparency, which seems to accomplish just that.

**Comment by snibgo** [comment]
> Ok, I spoke too soon...

Good stuff. I've already written the following, so I'll post it.

> However the animation becomes a choppy and stuttery, and seems to actually play slower.

Your delay is 4/100 seconds, so 25 frames per second (fps). This may be faster than your viewing computer can cope with. Try a longer delay, eg 6 or 10.

With "-layers optimize", the small file records each frame as just the smallest rectangle that contains just the pixels that have changed. A viewer may take advantage of this to update only those pixels on the screen, or it might take the slower route of building the entire frame in memory and copying that entire frame to the screen. This would use more CPU power, but has advantages of flexibility for zooming etc.

What software do you use to view the result GIF? I put the GIF as an image in a HTML page that I view with Firefox. Large GIFs (eg > 1m pixels per frame) often stutter in the first loop, but the second onwards is smooth, so there is caching somewhere. A specialist GIF viewer may be better.

Just for fun, you might try saving as a video format, eg .mp4, and viewing with a video player. This may, or may not, be more suitable for you.

> The duplicate frame worked for extending the last frame. How does it know it's the last frame of that layer to duplicate?

I used "+clone". This copies just the last image from the list as it was before the "(". I could have used "-clone -2--1" which would have copied the last two frames, or "-clone 0-2" for just the first 3 frames, and so on.

See the documentation at https://imagemagick.org/script/command-line-options.php#clone and other entries on that page. See also the animation pages linked from https://legacy.imagemagick.org/Usage/ .
